File: objc/TapstreamTest/Tapstream.m
#import "Tapstream.h"
#import "helpers.h"
#import "PlatformImpl.h"
#import "CoreListenerImpl.h"

@interface DelegateImpl : NSObject<Delegate> {}
- (int)getDelay;
- (bool)isRetryAllowed;
@end

@implementation DelegateImpl
- (int)getDelay
{
	return 0;
}

- (bool)isRetryAllowed
{
	return false;
}
@end




@interface Tapstream()

@property(nonatomic, STRONG_OR_RETAIN) id<Delegate> del;
@property(nonatomic, STRONG_OR_RETAIN) id<Platform> platform;
@property(nonatomic, STRONG_OR_RETAIN) id<CoreListener> listener;
@property(nonatomic, STRONG_OR_RETAIN) Core *core;

@end


@implementation Tapstream

@synthesize del, platform, listener, core;

- (id)initWithOperationQueue:(OperationQueue *)q accountName:(NSString *)accountName developerSecret:(NSString *)developerSecret hardware:(NSString *)hardware
{
	if((self = [super init]) != nil)
	{
		del = [[DelegateImpl alloc] init];
		platform = [[PlatformImpl alloc] init];
		listener = [[CoreListenerImpl alloc] initWithQueue:q];
		core = [[Core alloc] initWithDelegate:del
			platform:platform
			listener:listener
			accountName:accountName
			developerSecret:developerSecret
			hardware:hardware];
	}
	return self;
}

- (void)dealloc
{
	RELEASE(del);
	RELEASE(platform);
	RELEASE(listener);
	RELEASE(core);
	SUPER_DEALLOC;
}

- (void)fireEvent:(Event *)event
{
	[core fireEvent:event];
}

- (void)fireHit:(Hit *)hit completion:(void(^)(Response *))completion
{
	[core fireHit:hit completion:completion];
}

- (void)setResponseStatus:(int)status
{
	NSString *msg = [NSString stringWithFormat:@"Http %d", status];
	((PlatformImpl *)platform).response = AUTORELEASE([[Response alloc] initWithStatus:status message:msg]);
}

- (NSArray *)getSavedFiredList
{
	return ((PlatformImpl *)platform).savedFiredList;
}

- (int)getDelay
{
	return [core getDelay];
}

- (NSString *)getPostData
{
	return core.postData;
}

@end

File: cs/TapstreamTest/Tapstream.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Foundation;
using System.Runtime.InteropServices.WindowsRuntime;

namespace TapstreamMetrics.Sdk
{
    public sealed class Tapstream : Api
    {
        private class DelegateImpl : Delegate
        {
            private int delay = 0;

            public int GetDelay()
            {
                return delay;
            }

            public void SetDelay(int delay)
            {
                this.delay = delay;
            }

            public bool IsRetryAllowed()
            {
                return false;
            }
        }

        private Delegate del;
        private Platform platform;
        private CoreListener listener;
        private Core core;

        public Tapstream(OperationQueue queue, string accountName, string developerSecret, Config config)
        {
            del = new DelegateImpl();
            platform = new PlatformImpl();
            listener = new CoreListenerImpl(queue);
            core = new Core(del, platform, listener, accountName, developerSecret, config);
            core.Start();
        }

        public void FireEvent(Event e)
        {
            core.FireEvent(e);
        }

        public IAsyncOperation<Response> FireHitAsync(Hit h)
        {
            return core.FireHitAsync(h);
        }

        public void SetDelay(int delay)
        {
            del.SetDelay(delay);
        }

        public void SetResponseStatus(int status)
        {
            ((PlatformImpl)platform).response = new Response(status, String.Format("Http %d", status));
        }

        public string[] GetSavedFiredList()
        {
            if (((PlatformImpl)platform).savedFiredList == null)
            {
                return new string[] {};
            }
            return ((PlatformImpl)platform).savedFiredList.ToArray();
        }

        public int GetDelay()
        {
            return core.GetDelay();
        }

        public string GetPostData()
        {
            return core.GetPostData();
        }
    }
}
